aexpect/ops.py

```diff
@@ -0,0 +1,479 @@
+    :returns: whether the regular file exists
+    :rtype: bool
+    """
+    return test(session, path, flag="-f") == 0
+
+
+def is_nonzero_size_file(session, path):
+    """
+    Check if a regular file exists on a given guest.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: path to the regular file to check
+    :returns: whether the regular file exists and is nonzero size
+    :rtype: bool
+    """
+    return test(session, path, flag="-s") == 0
+
+
+###############################################################################
+# file ops
+###############################################################################
+
+
+def hash_file(session, filename, method='md5'):
+    """
+    Calculate hash of given file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file that should be hashed
+    :param str method: method for hashing
+    :returns: hash of file, a hex number
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if hash command failed (e.g. file not
+             found) or resulting output does not have expected format
+    :raises: :py:class:`ValueError` if given method is not (yet) supported
+    """
+    if method not in ['md5']:
+        raise ValueError('only method "md5" supported yet')
+
+    cmd = 'md5sum'
+    expect_len = 32
+
+    # run cmd on shell
+    status, output = session.cmd_status_output(cmd + ' ' + quote(filename))
+    if output:
+        output = output.strip()
+    if status != 0 or not output:
+        raise RuntimeError(f'Could not hash {filename} using {cmd}: {output}')
+
+    # parse output
+    hash_str = output.split(maxsplit=1)[0].lower()
+
+    # check length and all chars are hex
+    if expect_len and len(hash_str) != expect_len:
+        raise RuntimeError(f'Resulting hash string has unexpected length '
+                           f'{len(hash_str)}: {hash_str}')
+    if hash_str.strip('0123456789abcdef'):
+        raise RuntimeError('Resulting hash string has unexpected characters: '
+                           + hash_str)
+    return hash_str
+
+
+def extract_tarball(session, tarball, target_dir):
+    """
+    Extract tarball to given dir from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str tarball: full path of tar file that should be extracted
+    :param str target_dir: name of directory where tarball should be extracted
+    :raises: :py:class:`RuntimeError` if tar command returned non-null
+    """
+    cmd = f'tar -C {quote(target_dir)} --strip-components=1 -xapf {quote(tarball)}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to extract {tarball} to {target_dir}: {output}')
+
+
+def ls(session, dir_name):  # pylint: disable=C0103
```

Comment:
This is close to what I wrote at the beginning. IMO the basic API should be simple to use and cover the most common usage, while it would be nice to allow people to tweak the way they want. Here I'd expect: `ls(session, target=None, args='-1UNq')`. This way people could use `ls(session)` as well as completely change the view by `ls(session, '/foo', '--color')` and skip the default arguments. (note I do not say it's the best solution or that you have to adopt it everywhere. But for the raw commands it'd make sense to me)

Reply:
Definitely for the mechanism approach as well as adding as much flexibility of usage at an earlier stage as possible. For some more context that you both probably figured out by now: Many of these functions originate at different times (even different years) and implement only as much as was actually needed at the time of writing them in order to keep costs from extra development low (imagining these with tens of other tasks to cross of at the same time).

Even now contributing and merging these would result in a net positive effect for reuse by others and then based on incrementalism the ones that are of greater importance will probably see further extensions and improved applicability. This being said of course I don't mind generalizing some here and now especially if this only requires some minutes development time and just a few small extra steps. Will also make sure to retest the complete module with all our available tests once it stabilizes from a review point of view.